File: projects/10-multi-runtime-api-benchmark/lambdas/python/src/utils/metrics.py
import os
import time
import json
from typing import Dict, Any
from aws_lambda_powertools import Logger

logger = Logger()


class MetricsCollector:
    """Collect and report runtime metrics"""

    def __init__(self):
        self.runtime_name = os.environ.get('RUNTIME_NAME', 'python')
        self.start_time = time.time()
        self.cold_start = self._detect_cold_start()

    @staticmethod
    def _detect_cold_start() -> bool:
        """Detect if this is a cold start"""
        # Simple detection: check if global variable exists
        # In production, use Lambda Powertools or custom logic
        global _warm_start
        try:
            _warm_start
            return False
        except NameError:
            _warm_start = True
            return True

    def get_metrics(self) -> Dict[str, Any]:
        """Get current runtime metrics"""
        import psutil
        import sys

        process = psutil.Process()
        memory_info = process.memory_info()

        metrics = {
            'runtime': self.runtime_name,
            'cold_start': self.cold_start,
            'uptime_seconds': time.time() - self.start_time,
            'memory': {
                'rss_mb': memory_info.rss / 1024 / 1024,  # Resident Set Size
                'vms_mb': memory_info.vms / 1024 / 1024,  # Virtual Memory Size
            },
            'python_version': sys.version,
            'environment': os.environ.get('ENVIRONMENT', 'dev'),
        }

        # Add Lambda-specific context if available
        if 'AWS_LAMBDA_FUNCTION_NAME' in os.environ:
            metrics['lambda'] = {
                'function_name': os.environ.get('AWS_LAMBDA_FUNCTION_NAME'),
                'function_version': os.environ.get('AWS_LAMBDA_FUNCTION_VERSION'),
                'memory_limit_mb': os.environ.get('AWS_LAMBDA_FUNCTION_MEMORY_SIZE'),
                'log_group': os.environ.get('AWS_LAMBDA_LOG_GROUP_NAME'),
                'log_stream': os.environ.get('AWS_LAMBDA_LOG_STREAM_NAME'),
            }

        logger.info(f"Collected metrics: {json.dumps(metrics, default=str)}")
        return metrics

File: projects/10-multi-runtime-api-benchmark/lambdas/typescript/src/utils/metrics.ts
/**
 * Metrics collector for TypeScript Lambda
 */

interface LambdaMetrics {
  runtime: string;
  cold_start: boolean;
  uptime_seconds: number;
  memory: {
    heap_used_mb: number;
    heap_total_mb: number;
    external_mb: number;
    rss_mb: number;
  };
  node_version: string;
  environment: string;
  lambda?: {
    function_name?: string;
    function_version?: string;
    memory_limit_mb?: string;
    log_group?: string;
    log_stream?: string;
  };
}

export class MetricsCollector {
  private runtimeName: string;
  private startTime: number;
  private coldStart: boolean;

  // Use a module-level variable to track warm starts
  private static warmStarted = false;

  constructor() {
    this.runtimeName = process.env.RUNTIME_NAME || 'typescript';
    this.startTime = Date.now();
    this.coldStart = !MetricsCollector.warmStarted;
    MetricsCollector.warmStarted = true;
  }

  getMetrics(): LambdaMetrics {
    const memoryUsage = process.memoryUsage();

    const metrics: LambdaMetrics = {
      runtime: this.runtimeName,
      cold_start: this.coldStart,
      uptime_seconds: (Date.now() - this.startTime) / 1000,
      memory: {
        heap_used_mb: memoryUsage.heapUsed / 1024 / 1024,
        heap_total_mb: memoryUsage.heapTotal / 1024 / 1024,
        external_mb: memoryUsage.external / 1024 / 1024,
        rss_mb: memoryUsage.rss / 1024 / 1024,
      },
      node_version: process.version,
      environment: process.env.ENVIRONMENT || 'dev',
    };

    // Add Lambda-specific context if available
    if (process.env.AWS_LAMBDA_FUNCTION_NAME) {
      metrics.lambda = {
        function_name: process.env.AWS_LAMBDA_FUNCTION_NAME,
        function_version: process.env.AWS_LAMBDA_FUNCTION_VERSION,
        memory_limit_mb: process.env.AWS_LAMBDA_FUNCTION_MEMORY_SIZE,
        log_group: process.env.AWS_LAMBDA_LOG_GROUP_NAME,
        log_stream: process.env.AWS_LAMBDA_LOG_STREAM_NAME,
      };
    }

    console.log('Collected metrics:', JSON.stringify(metrics));
    return metrics;
  }
}
